### PR TITLE
bp: Migrate peer recovery from translog to retention lease

### PR DESCRIPTION
**Pull request by mkleen**
This is a none-trivial backport (for me) therefore i moved it to a distinct pr. The in-between bps from the list will follow as a separate pr.
Based on https://github.com/elastic/elasticsearch/commit/b9fbc8dc748

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/IndexSettings.java [hunk 1]
```diff
@@ -649,6 +651,11 @@ public int getTranslogRetentionTotalFiles() {
         return INDEX_TRANSLOG_RETENTION_TOTAL_FILES_SETTING.get(getSettings());
     }
 
+    private static boolean shouldDisableTranslogRetention(Settings settings) {
+        return INDEX_SOFT_DELETES_SETTING.get(settings)
+            && IndexMetadata.SETTING_INDEX_VERSION_CREATED.get(settings).onOrAfter(Version.V_4_3_0);
```
Comment (mkleen) [inline server/src/main/java/org/elasticsearch/index/IndexSettings.java:656]: Soft deletes were enabled as default in 4.3.

File: server/src/test/java/org/elasticsearch/index/seqno/RetentionLeaseIT.java [hunk 2]
```diff
@@ -95,11 +97,21 @@ public void testRetentionLeasesSyncedOnAdd() throws Exception {
...
 
+        for (int i = 0; i < length; i++) {
+            final String id = randomFrom(currentRetentionLeases.keySet());
+            final CountDownLatch latch = new CountDownLatch(1);
+            primary.removeRetentionLease(id, countDownLatchListener(latch));
+            // simulate a peer recovery which locks the soft deletes policy on the primary
+            final Closeable retentionLock = randomBoolean() ? primary.acquireHistoryRetentionLock(Engine.HistorySource.INDEX) : () -> {};
+            currentRetentionLeases.remove(id);
+            latch.await();
+            retentionLock.close();
```
Comment (seut) [inline server/src/test/java/org/elasticsearch/index/seqno/RetentionLeaseIT.java:114]: why is that needed, where is this coming from?

**Review by seut — COMMENTED** [review]
Looks good, but added one question related to the tests.

File: server/src/test/java/org/elasticsearch/index/seqno/RetentionLeaseIT.java (hunk 2 above)
Comment (mkleen) [reply server/src/test/java/org/elasticsearch/index/seqno/RetentionLeaseIT.java:114]: It is part of the `testRetentionLeasesSyncedOnAdd` at the state of the backport https://github.com/crate/crate/blob/c54f405b817d51d97ba1b621e1f421ca9efccda7/server/src/test/java/org/elasticsearch/index/seqno/RetentionLeaseIT.java#L106 and it was introduced here https://github.com/elastic/elasticsearch/pull/37398.

Comment (seut) [reply server/src/test/java/org/elasticsearch/index/seqno/RetentionLeaseIT.java:114]: Afaik this code block is part of a missing `testRetentionLeaseSyncedOnRemove` test and should not be added inside the `testRetentionLeasesSyncedOnAdd` test method. See  https://github.com/elastic/elasticsearch/blob/b9fbc8dc74826329f0bb00fc4f531507a12ec327/server/src/test/java/org/elasticsearch/index/seqno/RetentionLeaseIT.java

Comment (mkleen) [reply server/src/test/java/org/elasticsearch/index/seqno/RetentionLeaseIT.java:114]: oh, you are right, my fault. thank you!

Comment (mkleen) [reply server/src/test/java/org/elasticsearch/index/seqno/RetentionLeaseIT.java:114]: I took the wrong diff.

**Review by seut — APPROVED** [review]
👍
